modules/transaction-payment/src/lib.rs

```diff
@@ -825,31 +821,101 @@ where
+		let fee_surplus: Result<Balance, DispatchError> = match call.is_sub_type() {
+			Some(Call::with_fee_path { fee_swap_path, .. }) => {
+				let supply_currency_id = *fee_swap_path.first().expect("should match a non native asset");
+				T::DEX::swap_with_specific_path(
+					who,
+					fee_swap_path,
+					SwapLimit::ExactTarget(
+						T::MultiCurrency::free_balance(supply_currency_id, who),
+						custom_fee_amount,
+					),
+				)
+				.map(|_| custom_fee_surplus)
+			}
+			Some(Call::with_fee_currency { currency_id, .. }) => {
+				Self::swap_from_pool_or_dex(who, custom_fee_amount, *currency_id).map(|_| custom_fee_surplus)
+			}
+			Some(_) | None => Self::native_then_alternative_or_default(who, fee, reason),
+		};
+		fee_surplus
+	}
 
-		// make sure add extra gap to keep alive after swap.
-		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
-		// native is not enough, try swap native from fee pool to pay fee and gap.
-		Self::swap_native_asset(who, amount);
+	fn ensure_can_charge_fee(who: &T::AccountId, fee: PalletBalanceOf<T>, reason: WithdrawReasons) {
+		let _ = Pallet::<T>::native_then_alternative_or_default(who, fee, reason);
 	}
 
-	/// Iterate order list, break if can swap out enough native asset amount with user's foreign
-	/// asset. make sure trading path is exist in dex, if the trading pair is not exist in dex, even
-	/// though we have setup it in charge fee pool, we can't charge fee with this foreign asset.
-	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
-		let native_currency_id = T::NativeCurrencyId::get();
-		for trading_path in Self::get_trading_path(who) {
-			if trading_path.last() == Some(&native_currency_id) {
-				let supply_currency_id = *trading_path.first().expect("should match a non native asset");
-				if T::MultiCurrency::free_balance(supply_currency_id, who).is_zero() {
-					continue;
+	/// If native is enough, do nothing, return `Ok(0)` means there are none extra surplus fee.
+	/// If native is not enough, try swap from tx fee pool or dex. As user can set his own
+	/// `AlternativeFeeSwapPath`, this will direct swap from dex. Sometimes, user setting of
+	/// `AlternativeFeeSwapPath` may be wrong or dex is not available, or user do not set any
+	/// `AlternativeFeeSwapPath`, then use the `DefaultFeeTokens` to swap from tx fee pool.
+	fn native_then_alternative_or_default(
+		who: &T::AccountId,
+		fee: PalletBalanceOf<T>,
+		reason: WithdrawReasons,
+	) -> Result<Balance, DispatchError> {
+		// native asset is not enough
+		if let Some(amount) = Self::native_is_enough(who, fee, reason) {
+			let fee_surplus = T::AlternativeFeeSurplus::get().mul_ceil(fee);
+			let fee_amount = fee_surplus.saturating_add(amount);
+
+			// alter native fee swap path, swap from dex: O(1)
+			if let Some(path) = AlternativeFeeSwapPath::<T>::get(who) {
+				let supply_currency_id = *path.first().expect("should match a non native asset");
```

Comment:
this shouldn't fail but still better to avoid expect here and just return a DispatchError